### ISSUE TITLE
"Used before assigned" error with arrow functions and strict null checks

### ISSUE THREAD

**Issue by Christoph-D** [Microsoft/TypeScript#9382]
**TypeScript Version:** 

1.9.0-dev.20160627-1.0

**Code**

```ts
const f = () => { return f; };
```
Compiled with `--strictNullChecks`.

**Expected behavior:**

No errors.

**Actual behavior:** 

```
foo.ts(1,26): error TS2454: Variable 'f' is used before being assigned.
```

It sems to be impossible to call `f` before `f` is assigned, so I don't think this should trigger an error.

**Comment by mhegazy** [comment]
The error message is a bit misleading. the message is about the type system trying to know what is the type of `f` and catching its own tail. so an explicit type annotation would resolve the issue. so:

```ts
const f: any = () => { return f; };
```

**Comment by Christoph-D** [comment]
This gives the same error, although I added an explicit type annotation.
```ts
const f: () => void = () => { f; }
```
But the following gives no error:
```ts
let f: () => void;
f = () => { f; }
```

**Comment by bolismauro** [comment]
@Christoph-D I don't receive the error if I change the definition of the function from `const` to `let`.
So basically this gives me an error
```js
const t2 = (): number => 1 + t2();
```

while this does not
```js
let t = (): number => 1 + t();
```

**Comment by mhegazy** [comment]
should be fixed by https://github.com/Microsoft/TypeScript/pull/10815
